### DISCUSSION TITLE
Accepting all open recommendations (in a sentence)

### DISCUSSION THREAD

**Discussion by senier** [inception-project/inception#3902] [Q&A] [answered]
Hi all,

first of all: thanks for creating INCEpTION and making it available as open source software! I love how easy you get started, but at the same time how powerful the tool is.

I have a usage questions: When using a recommender, suggestions can be accepted by single-clicking. Once the recommender gets pretty accurate, this can become a lot of clicks just for accepting all the correct suggestions. A workflow where you modify / delete incorrect recommendations first and then accept all the other ones in a sentence by a single key combination could boost my efficiency. Is that possible?

Thanks!

Cheers,
Alex

**Comment by reckart** [comment, marked as answer]
In the recommender sidebar, there is a button `Accept best` which may work for you. If you are working with multiple recommenders, there will be one of these buttons for each recommender though.

Also note that in recent 27.x versions, the annotation sidebar offers more convenient access to recommendations including the ability to directly accept/reject them from the sidebar.

**Reply by senier** [reply]
Thanks for you quick reply! I'm using 27.2, but I cannot spot an "Accept best" button. I didn't mention that I'm using a remote recommender - does that make a difference?

I noticed the new Accept/Reject buttons in the annotations sidebar. It seems like it always shows all labels for the whole document. Is there a way to limit it to the currently displayed sentences?

![image](https://user-images.githubusercontent.com/6357197/228352776-2926d18a-c770-46d7-ae23-ba0d5cb81d3d.png)

**Reply by reckart** [reply]
Hm - that button not being there looks like a bug then... thanks for pointing this out.

The annotation sidebar is for the whole document. It is also a tool for navigating within the document. Why would you want to limit it just to a sentence?

**Reply by senier** [reply]
Right now, when I label data, I make sure a sentence is labeled correctly (i.e. wrong/missing labels corrected and correct labels accepted) before going to the next sentence. I have the UI configured to show only a single sentence. Thinking about it, I could just leave correct recommendations until I'm done with the whole document and then accept all suggestions. Is that a typical workflow? Could it happen that recommendations I saw earlier change to something unintended? Or are recommendations fixed once I've seen them?

**Reply by reckart** [reply]
If you are working with a non-trainable recommender, then the recommendations are fetched only once on document access and remain unchanged thereafter. If the recommender is deterministic based on the document text and not sensitive to annotations, then you should always get the same recommendations for it in every session.

If you are working with trainable recommenders, then recommendations can change as you annotate. Every time you create/accept or change an annotation, a new training run is scheduled.

If you are working one sentence at a time, you might consider splitting your data up such that every document contains exactly one sentence. That would allow you to mark each sentence as finished as you go along so you can keep better track of progress. Or alternatively put a constant number of sentences into each document (e.g. 10) - like a batch. You might then consider to also display all of these 10 sentences at once in the editor to make sure that you do not accidentally miss something before marking the document/batch as finished.

**Reply by senier** [reply]
Thanks for the explanation! Sounds like the "Accept best" approach might work for me (once the issue is fixed). Adding an "Accept all recommendations in the current view" feature might still be a reasonable enhancement. It would give the user the flexibility to change all invalid recommendations in the current window (be that 1 or 10 sentences) without having to structure the input data in a particular way or worry that a trainable recommender changes correct suggestions later.

Thanks again for your time and quick response.

**Reply by reckart** [reply]
Are the documents you are using actual texts consisting of multiple sentences, or is each document just a random collection of sentences?

**Reply by senier** [reply]
The latter. They are mostly independent, not part of a bigger text.

**Reply by reckart** [reply]
In that case, are you not arranging your input data already anyway? Wouldn't it be little extra work to arrange it slightly differently?

**Reply by reckart** [reply]
Btw. here is the bug report you can subscribe to for notifications: https://github.com/inception-project/inception/issues/3909

**Reply by senier** [reply]
Sure, I've got a script that converts the data anyway which I could easily change to produce smaller chunks. I was just thinking that such a feature might still be useful to some :smiley: 

I saw the ticket, thanks!

**Comment by reckart** [comment]
Note though that the recommender only learns from accepted annotations. Thus, if your workflow consists only of rejecting suggestions, then the recommender will not improve.

**Reply by senier** [reply]
Good to know. I'm only using a non-trainable remote recommender based on a spacy model. So I guess this is of no concern for me right now.
